pilot/pkg/xds/eds.go

```diff
@@ -462,11 +474,19 @@ func (eds *EdsGenerator) buildEndpoints(proxy *model.Proxy,
 			}
 		}
 		builder := NewEndpointBuilder(clusterName, proxy, req.Push)
-		if marshalledEndpoint, f := eds.Server.Cache.Get(builder); f && !features.EnableUnsafeAssertions {
-			// We skip cache if assertions are enabled, so that the cache will assert our eviction logic is correct
-			resources = append(resources, marshalledEndpoint)
-			cached++
-		} else {
+
+		marshalledEndpoint, token := eds.Server.Cache.Get(&builder)
+		if marshalledEndpoint != nil && !features.EnableUnsafeAssertions {
+			// If PA does not change or it has changed but the eds cache has already been refreshed in this period.
+			if useCache || (!useCache && token >= model.CacheToken(req.Start.UnixNano())) {
```

Comment:
this is a good idea but leaks internal details to caller. Maybe we can make a function on the cache like GetRequest(builder, req) that encapsulates these details in the cache?

I would say it could even do the type check (Peer Auth ) but that may be slow to do it in the inner loop

Reply:
>Maybe we can make a function on the cache like GetRequest(builder, req) that encapsulates these details in the cache?

make sense, will do that